### PR TITLE
[github-actions] add gcc-12 to build matrix and fix warnings

### LINKED ISSUE #8117: openthread/src/core/common/message.hpp:122:38: error: ‘nonnull’ argument ‘aMessage’ compared to NULL [-Werror=nonnull-compare]
**Describe the bug** A clear and concise description of what the bug is.

Build failure with gcc 12.2.0

**To Reproduce** Information to reproduce the behavior, including:

1. Git commit id: 30d65bca4896a38e0e8a9e0b3cb0b9d458454ef2
2. IEEE 802.15.4 hardware platform: N/A
3. Build steps: `./bootstrap && ./script/cmake-build posix -DOT_SPINEL_RESET_CONNECTION=ON -DOT_THREAD_VERSION=1.2 -DOT_DAEMON=ON`

4. Network topology: N/A

**Expected behavior** A clear and concise description of what you expected to happen.

Build completes without warnings or errors

**Console/log output** If applicable, add console/log output to help explain your problem.

```
In file included from /home/zyga/projects/openthread/src/core/crypto/aes_ccm.hpp:43,
                 from /home/zyga/projects/openthread/src/core/thread/mle.hpp:45,
                 from /home/zyga/projects/openthread/src/core/thread/mle.cpp:34:
/home/zyga/projects/openthread/src/core/thread/mle.cpp: In member function ‘void ot::Mle::Mle::SendDelayedResponse(TxMessage&, const DelayedResponseMetadata&)’:
/home/zyga/projects/openthread/src/core/common/message.hpp:122:38: error: ‘nonnull’ argument ‘aMessage’ compared to NULL [-Werror=nonnull-compare]
  122 |         if (((aError) != kErrorNone) && ((aMessage) != nullptr)) \
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/zyga/projects/openthread/src/core/thread/mle.cpp:1643:5: note: in expansion of macro ‘FreeMessageOnError’
 1643 |     FreeMessageOnError(&aMessage, error);
      |     ^~~~~~~~~~~~~~~~~~
```

**Additional context** Add any other context about the problem here.

```
zyga@lambert:~/projects/openthread$ gcc --version
gcc (Ubuntu 12.2.0-1ubuntu1) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

zyga@lambert:~/projects/openthread$ cmake --version
cmake version 3.24.1

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

### PR DESCRIPTION
**Pull request by jwhui**
Fixes #8117

### CODE REVIEW

File: src/ncp/ncp_hdlc.hpp [hunk 1]
```diff
@@ -128,9 +128,9 @@ class NcpHdlc : public NcpBase
                                                       Spinel::Buffer *         aBuffer);
     otNcpHdlcSendCallback mSendCallback;
 
+    Hdlc::FrameBuffer<kHdlcTxBufferSize> mHdlcBuffer;
```
Comment (zyga) [inline src/ncp/ncp_hdlc.hpp:131]: Thank you for fixing this. I didn't report it but it was next on the list :)

**Comment by zyga** [comment]
This looks good. I don't know how much you care about supporting/testing Ubuntu 20.04. If the cost is not too high it might be better to keep testing against it for some time, and to drop it later.

File: src/core/net/dnssd_server.cpp [hunk 2]
```diff
@@ -219,7 +219,11 @@ void Server::SendResponse(Header                  aHeader,
 
     error = aSocket.SendTo(aMessage, aMessageInfo);
 
-    FreeMessageOnError(&aMessage, error);
+    // do not use FreeMessageOnError() to avoid null check on nonnull pointer
+    if (error != kErrorNone)
+    {
+        aMessage.Free();
+    }
```
Comment (abtink) [inline src/core/net/dnssd_server.cpp:226]: Can combine with `if` below?

**Review by abtink — APPROVED** [review]
LGTM. Thanks.

**Comment by jwhui** [comment]
> This looks good. I don't know how much you care about supporting/testing Ubuntu 20.04. If the cost is not too high it might be better to keep testing against it for some time, and to drop it later.

Good thought. However, I believe we are using 20.04 in other GitHub Actions, so it is OK to remove these. If we find this to be an issue in the future, we can add them back in.

**Pull request merged**